server/plugins/datasources/mongodb/index.ts

```diff
@@ -22,6 +22,23 @@ export default class MongodbService implements QueryService {
         case 'insert_many':
           result = await db.collection(queryOptions.collection).insertMany(JSON.parse(queryOptions.documents));
           break;
+        case 'find_one':
+          result = await db.collection(queryOptions.collection).findOne(JSON.parse(queryOptions['query']));
+          break;
+        case 'find':
+          let resultQueryBuilder = db.collection(queryOptions.collection)
+          if ("query" in queryOptions && queryOptions['query'])
+            resultQueryBuilder = resultQueryBuilder.find(JSON.parse(queryOptions['query']))
+          else
+            resultQueryBuilder = resultQueryBuilder.find({})
```

Comment:
_(Sorry for chiming in abruptly here, was working on a similar feature locally while I was evaluating tooljet for internal use at my org and came across this thread)_

Did this work for you locally on large find queries?

Pretty sure what you need here is `resultQueryBuilder.find({}).toArray()` given that the upstream service is [awaiting the return value](https://github.com/ToolJet/ToolJet/blob/6acc4b43ab663a885a2498e966074e37e39a4300/server/src/services/data_queries.service.ts#L81). `find()` itself just[ returns a cursor](https://mongodb.github.io/node-mongodb-native/4.0/classes/collection.html#find) so that upstream expression will yield no actual documents.